modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt

```diff
@@ -142,33 +220,64 @@ class VaultWatcherService(private val tokenObserver: TokenObserver, private val
+            { token ->
+                predicate(token) && token.state.data.issuer == issuer
+            }
+        } else {
+            predicate
+        })
+
         val lockedTokens = mutableListOf<StateAndRef<FungibleToken>>()
-        var amountLocked: Amount<TokenType> = requiredAmount.copy(quantity = 0)
-        for (tokenStateAndRef in bucket.keys) {
+        val bucket: Iterable<StateAndRef<FungibleToken>> = if (owner is Holder.TokenOnly) {
+            val currentPredicate = enrichedPredicate.get()
+            //why do we do this? It doesn't really make sense to index on token type, as it's very likely that there will be very few types of tokens in a given vault
+            //so instead of relying on an indexed view, we can create a predicate on the fly which will constrain the selection to the correct token type
+            //we will revisit in future if this assumption turns out to be wrong
+            enrichedPredicate.set {
+                val stateTokenType = it.state.data.tokenType
+                currentPredicate(it) &&
+                        stateTokenType.fractionDigits == requiredAmount.token.fractionDigits &&
+                        requiredAmount.token.tokenClass == stateTokenType.tokenClass &&
+                        requiredAmount.token.tokenIdentifier == stateTokenType.tokenIdentifier
+            }
+            __backingMap.keys
+        } else {
+            val indexedView = getOrCreateIndexViewForHolderType(owner.javaClass)
+            getTokenBucket(owner, requiredAmount.token.tokenClass, requiredAmount.token.tokenIdentifier, indexedView)
+        }
+
+        val requiredAmountWithoutIssuer = requiredAmount.withoutIssuer()
+        var amountLocked: Amount<TokenType> = requiredAmountWithoutIssuer.copy(quantity = 0)
+        val finalPredicate = enrichedPredicate.get()
+        for (tokenStateAndRef in bucket) {
             // Does the token satisfy the (optional) predicate eg. issuer?
-            if (predicate.invoke(tokenStateAndRef)) {
+            if (finalPredicate.invoke(tokenStateAndRef)) {
                 // if so, race to lock the token, expected oldValue = PLACE_HOLDER
-                if (bucket.replace(tokenStateAndRef, PLACE_HOLDER, selectionId)) {
+                if (__backingMap.replace(tokenStateAndRef, PLACE_HOLDER, selectionId)) {
                     // we won the race to lock this token
                     lockedTokens.add(tokenStateAndRef)
                     val token = tokenStateAndRef.state.data
                     amountLocked += uncheckedCast(token.amount.withoutIssuer())
-                    if (amountLocked >= requiredAmount) {
+                    if (amountLocked >= requiredAmountWithoutIssuer) {
                         break
                     }
                 }
             }
         }
 
-        if (!allowShortfall && amountLocked < requiredAmount) {
+        if (!allowShortfall && amountLocked < requiredAmountWithoutIssuer) {
```

Comment:
basically allows you to select "all" tokens in a bucket without having to guess exactly the amount available - very useful for testing.

Reply:
OK got you. maybe we can add an overload for select tokens which is only visible for testing then?

Reply:
yeah unfortunately - it was released in 1.0. 

It's not just useful for testing, it would also quickly allow you to select "all" tokens of a given type if you had to calculate a balance